File: Y1S2/ESC190/lab07/lab07.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


typedef struct CircularQueue{
    int size;
    int *data;
    int start;
    int end;
}CircularQueue;

void initialize(CircularQueue *q, int size) {
    q->data = (int *)malloc(size * sizeof(int));
    for (int i =0;i<size;i++){
        q->data[i]=0;
    }
    q->size = size;
    q->start = size/2-1;
    q->end = size/2-1;
}

void enqueue(CircularQueue *q, int item) {
    q->end = (q->end + 1) % q->size;
    if (q->data[q->end]!=0){
        q->data = (int *)realloc(q->data, (q->size+1) * sizeof(int));
        memmove(q->data+q->start+1, q->data+q->start, sizeof(int)*(q->size-q->start));
        q->size++;
        if (q->start>q->end){
            q->start++;
        }
    }
    q->data[q->end] = item;
}

void dequeue(CircularQueue *q) {
    q->start = (q->start + 1) % q->size;
    q->data[q->start] = 0;
}

void print(CircularQueue *q){
    for (int i = 0;i<q->size;i++){
        printf("%d ", q->data[i]);
    }
    printf("\n");
}

void destroyQueue(CircularQueue *q) {
    free(q->data);
    free(q);
}


int main() {
    CircularQueue q;
    initialize(&q, 5);
    enqueue(&q, 1);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    dequeue(&q);
    print(&q);
    dequeue(&q);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    enqueue(&q, 1);
    print(&q);
    dequeue(&q);
    print(&q);
    dequeue(&q);
    print(&q);
    dequeue(&q);
    print(&q);
    dequeue(&q);
    print(&q);
    return 0;
}

File: Y1S2/ESC190/lab07/lab07.py
class CircularQueue:
    def __init__(self, size):
        self.size = size
        self.data = [None]*size
        self.start = size//2-1
        self.end = size//2-1
    def enqueue(self, item):
        self.end=(self.end+1)%self.size
        self.data[self.end]=item
        print(self)

    def dequeue(self):
        self.start=(self.start+1)%self.size
        self.data[self.start]=None
        print(self)

    def __str__(self):
        out = ""
        for value in self.data:
            out = out +str(value)+" "
        out = out+"\n"+str(self.start)+", "+str(self.end)
        return out

    def __lt__(self, other):
        for i in range(min((self.end-self.start)%self.size, (other.end-other.start)%other.size)):
            print(i)
            if self.data[(self.start + 1+ i) % self.size] < other.data[(other.start +1+ i) % other.size]:
                return True
            elif self.data[(self.start +1+ i) % self.size] > other.data[(other.start +1+ i) % other.size]:
                return False
        return True

# Question 1
cq = CircularQueue(5)
cq.enqueue(1)
cq.enqueue(1)
cq.enqueue(1)
cq.enqueue(1)
cq.enqueue(1)
cq.dequeue()
cq.dequeue()
cq.dequeue()
cq.dequeue()
cq.dequeue()



# Question 2
cq1 = CircularQueue(6)
cq1.enqueue(2)
cq1.enqueue(5)
cq1.enqueue(10)
cq1.enqueue(11)

cq2 = CircularQueue(6)
cq2.enqueue(2)
cq2.enqueue(5)
cq2.enqueue(10)
cq2.enqueue(10)

cq3 = CircularQueue(6)
cq3.enqueue(3)
cq3.enqueue(5)
cq3.enqueue(10)

queues = [cq1, cq2, cq3]
sorted_queues = sorted(queues)

print("\nSorted Queues:")
for cq in sorted_queues:
    print(cq)
